util/secscan/v4/api.py

```diff
@@ -0,0 +1,203 @@
+    the expected schema.
+    """
+
+
+@add_metaclass(ABCMeta)
+class SecurityScannerAPIInterface(object):
+    @abstractmethod
+    def state(self):
+        """
+        The state endpoint returns a json structure indicating the indexer's internal configuration state. 
+        A client may be interested in this as a signal that manifests may need to be re-indexed.
+        """
+        pass
+
+    @abstractmethod
+    def index(self, manifest, layers):
+        """
+        By submitting a Manifest object to this endpoint Clair will fetch the layers, 
+        scan each layer's contents, and provide an index of discovered packages, repository and distribution information.
+        Returns a tuple of the `IndexReport` and the indexer state.
+        """
+        pass
+
+    @abstractmethod
+    def index_report(self, manifest_hash):
+        """
+        Given a Manifest's content addressable hash an `IndexReport` will be retrieved if exists.
+        """
+        pass
+
+    @abstractmethod
+    def vulnerability_report(self, manifest_hash):
+        """
+        Given a Manifest's content addressable hash a `VulnerabilityReport` will be created. 
+        The Manifest must have been Indexed first via the Index endpoint.
+        """
+        pass
+
+
+Action = namedtuple("Action", ["name", "payload"])
+
+actions = {
+    "State": lambda: Action("State", ("GET", "state", None)),
+    "Index": lambda manifest: Action("Index", ("POST", "index_report", manifest)),
+    "GetIndexReport": lambda manifest_hash: Action(
+        "GetIndexReport", ("GET", "index_report/" + manifest_hash, None)
+    ),
+    "GetVulnerabilityReport": lambda manifest_hash: Action(
+        "GetVulnerabilityReport", ("GET", "vulnerability_report/" + manifest_hash, None,)
+    ),
+}
+
+
+class ClairSecurityScannerAPI(SecurityScannerAPIInterface):
+    def __init__(self, endpoint, client, storage):
+        self._client = client
+        self._storage = storage
+        self.secscan_api_endpoint = urljoin(endpoint, "/api/v1/")
+
+    def state(self):
+        try:
+            resp = self._perform(actions["State"]())
+        except Non200ResponseException as ex:
+            msg = (
+                "Security scanner endpoint responded with non-200 HTTP status code: %s"
+                % ex.message.status_code
+            )
+            logger.exception(msg)
+            raise Exception(msg)
+        except requests.exceptions.ConnectionError as ce:
+            logger.exception("Connection error when trying to connect to security scanner endpoint")
+            msg = (
+                "Connection error when trying to connect to security scanner endpoint: %s"
+                % ce.message
+            )
+            raise Exception(msg)
+
+        return resp.json()
+
+    def index(self, manifest, layers):
+        assert isinstance(manifest, ManifestDataType) and not manifest.is_manifest_list
```

Comment:
Add a TODO for us either to add a field to the manifest about whether it is security indexable or to define a whitelist of content types somewhere